wire-ios-request-strategy/Sources/Message Sending/MessageSender.swift

```diff
@@ -0,0 +1,240 @@
+        }
+    }
+}
+
+enum MessageTarget {
+    case Users(usersIds: [QualifiedID])
+    case Clients(recipients: [QualifiedClientID])
+    case Conversation(usersToIgnore: Set<QualifiedID>)
+}
+
+enum MessageSendError: Error {
+    case messageProtocolMissing
+    case unresolvedApiVersion
+    case failedToGenerateRequest
+    case failedToParseResponse
+    case gaveUpRetrying
+}
+
+typealias Message = ProteusMessage & MLSMessage
+
+public class MessageSender {
+
+    public init (
+        httpClient: HttpClient,
+        clientRegistrationDelegate: ClientRegistrationDelegate,
+        sessionEstablisher: SessionEstablisher,
+        context: NSManagedObjectContext
+    ) {
+        self.httpClient = httpClient
+        self.clientRegistrationDelegate = clientRegistrationDelegate
+        self.sessionEstablisher = sessionEstablisher
+        self.managedObjectContext = context
+    }
+
+    private let httpClient: HttpClient
+    private let managedObjectContext: NSManagedObjectContext
+    private let clientRegistrationDelegate: ClientRegistrationDelegate
+    private let sessionEstablisher: SessionEstablisher
+    private let requestFactory = ClientMessageRequestFactory()
+    private let processor = MessageSendingStatusPayloadProcessor()
+
+    func sendMessage(message: any Message) async -> Swift.Result<Void, MessageSendError> {
+        // FIXME: [jacob] wait for message dependencies to resolve
+
+        return await attemptToSend(message: message)
+    }
+
+    private func attemptToSend(message: any Message) async -> Swift.Result<Void, MessageSendError> {
+        guard let apiVersion = BackendInfo.apiVersion else { return .failure(MessageSendError.unresolvedApiVersion)}
+        guard let messageProtocol = managedObjectContext.performAndWait({
+            message.conversation?.messageProtocol
+        }) else {
+            return .failure(MessageSendError.messageProtocolMissing)
+        }
+
+        return switch messageProtocol {
+        case .proteus:
+            await attemptToSendWithProteus(message: message, apiVersion: apiVersion)
+
+        case .mls:
+            await attemptToSendWithMLS(message: message, apiVersion: apiVersion)
+        }
+    }
+
+    private func attemptToSendWithProteus(message: any ProteusMessage, apiVersion: APIVersion) async -> Swift.Result<Void, MessageSendError> {
+        guard let conversation = managedObjectContext.performAndWait({
+            message.conversation
+        }) else {
+            return .failure(MessageSendError.messageProtocolMissing)
+        }
+        guard let request = managedObjectContext.performAndWait({
+            requestFactory.upstreamRequestForMessage(message, in: conversation, apiVersion: apiVersion)
+        }) else {
+            return .failure(MessageSendError.failedToGenerateRequest)
+        }
+
+        let response = await httpClient.send(request)
+
+        return if response.result == .success {
```

Comment:
I wasn't aware you could return an if here, can we?

Reply:
It was added quite recently to swift